src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts

```diff
@@ -102,7 +105,13 @@ export class TextModelTreeSitter extends Disposable {
 	}
 
 	private async _onDidChangeContent(treeSitterTree: TreeSitterParseResult, changes: IModelContentChange[]) {
-		return treeSitterTree.onDidChangeContent(this.model, changes);
+		const oldTree = treeSitterTree.tree?.copy();
+		await treeSitterTree.onDidChangeContent(this.model, changes);
+		if (oldTree && treeSitterTree.tree) {
+			const diff = oldTree.getChangedRanges(treeSitterTree.tree);
```

Comment:
This `diff` propagates all the way up to the tokenizer.
However, I'm not sure how "local" the tokenization rules are.
Lets say you have this class:

```ts
class Foo {
	public bar() {}
}
```

Now, you change `public` to `private`:
```ts
class Foo {
	private bar() {}
}
```

In the tree, the keyword `public` now changed to `private`, so that's (hopefully) the range you'd get.
However, it might not be sufficient to only re-render the `private` token now, as there might be a rule which makes `bar` bold depending on if the method is public or private.

Reply:
Just tested with your example, but I put `bar() {}` on a separate line. As far as I can see, `getChangedRanges` will return ranges that include `bar() {}` as part of the changed ranges.